drivers/pio/sunxi-pio.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <bitmap.h>
+#include <dm.h>
+#include <error.h>
+#include <mmio.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <util.h>
+#include <drivers/clock.h>
+#include <drivers/irqchip.h>
+#include <drivers/pio.h>
+#include <drivers/pio/sunxi-pio.h>
+
+#define MAX_PORTS         8
+#define MAX_PINS_PER_PORT 32
+
+#define PIN_INDEX(pin)    ((pin) & BITMASK(0, 5))
+#define PIN_PORT(pin)     ((pin) >> 5)
+
+#define PIN_MODE(mode)    ((mode) & BITMASK(0, 3))
+
+#define CONFIG_REG(pin) \
+	(0x0000 + PIN_PORT(pin) * 0x24 + (PIN_INDEX(pin) >> 3) * 4)
+#define DATA_REG(pin) \
+	(0x0010 + PIN_PORT(pin) * 0x24)
+
+#define CONFIG_OFFSET(pin)    ((PIN_INDEX(pin) & BITMASK(0, 3)) * 4)
+
+#define INT_CONTROL_REG(port) (0x0210 + (port) * 0x20)
+#define INT_STATUS_REG(port)  (0x0214 + (port) * 0x20)
+
+static bool
+sunxi_pio_read_pin(struct device *dev, uint8_t pin)
+{
+	return mmio_read32((dev->regs + DATA_REG(pin)) & BIT(PIN_INDEX(pin)));
```

Comment:
The `&` goes *outside* the call to `mmio_read32`, so you mask the *value* and not the *address* (I've done this before and it leads to really really nasty bugs).

Reply:
Good catch. Fixed.